File: LeetCode/python3/15.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        print(nums)
        i = 0
        while i < len(nums):
            target = -nums[i]
            x, y = i + 1, len(nums) - 1
            while x < y:
                if nums[x] + nums[y] < target:
                    x += 1
                elif nums[x] + nums[y] > target:
                    y -= 1
                else:
                    X = nums[x]
                    Y = nums[y]
                    res.append([nums[i], X, Y])
                    while x < y and nums[x] == X:
                        x += 1
                    while x < y and nums[y] == Y:
                        y -= 1

            while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                i += 1
            i += 1
        return res

File: LeetCode/cpp/15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int> &nums) {
        vector<vector<int>> ret;
        auto size = nums.size();
        std::sort(nums.begin(), nums.end());
        for (int j = 0; j < size; ++j) {
            int &&target = -nums[j];
            int x = j + 1, y = size - 1;
            while (x < y) {
                if (nums[x] + nums[y] < target)
                    ++x;
                else if (nums[x] + nums[y] > target)
                    --y;
                else {
                    ret.push_back({nums[j], nums[x], nums[y]});
                    do
                        ++x;
                    while (x < y && nums[x] == nums[x - 1]);
                    do
                        --y;
                    while (x < y && nums[y] == nums[y + 1]);
                }
            }
            while (j + 1 < size && nums[j + 1] == nums[j])
                ++j;
        }
        return ret;
    }
};
